nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -363,6 +354,54 @@ impl<'a> BlueprintBuilder<'a> {
         })
     }
 
+    // Helper method to create a `BuilderExternalNetworking` allocator at the
+    // point of first use. This is to work around some timing issues between the
+    // planner and the builder: `BuilderExternalNetworking` wants to know what
+    // resources are in use by running zones, but the planner constructs the
+    // `BlueprintBuilder` before it expunges zones. We want to wait to look at
+    // resources are in use until after that expungement happens; this
+    // implicitly does that by virtue of knowing that the planner won't try to
+    // allocate new external networking until after it's done expunging and has
+    // started to provision new zones.
```

Comment:
This is all fine, but just to be clear: is this _just_ trying to avoid an extra planning lap when expunging a zone before you can use its resources again?  I feel like there's another issue here and I also feel like we talked about this a bunch -- sorry -- but I am not now remembering why else this is important.

Reply:
In principle, this is just trying to avoid an extra planning lap, yes. In practice, the way the planner is currently written, it expects to be able to reuse resources within a single planning execution and fails if it cannot:

1. It first performs any necessary zone expungements (based on sled or disk expungements from `PlanningInput`)
2. It then adds any services as required by the policy, _taking into account_ any services that are now underrepresented because of expungements that happened in step 1

Without these helpers to delay the construction of the allocators, we were seeing this:

1. Allocators are constructed. They assume resource availability based solely on the parent blueprint.
2. Planner performs expungement. Suppose we expunge an external DNS zone.
3. Planner performs new service addition. It sees we're down an external DNS zone, so tries to add one. This fails, because the allocator was based on the parent blueprint, in which all the external DNS IPs were in use by in-service zones.

Maaaybe another way to tackle this would be to make the planner bail out if it performed any expungements? But that feels like a much bigger change.

Reply:
Either way's fine with me but maybe we could add this context to the comment?